File: examples/go/table/agg/handlers/join.go
package handlers

import (
	"fmt"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardJoinTable(state TableState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Table does not exist")
	}
	return nil
}

func validateJoinTable(cmd *examples.JoinTable, state TableState) (int32, error) {
	if len(cmd.PlayerRoot) == 0 {
		return 0, angzarr.NewCommandRejectedError("player_root is required")
	}

	if state.FindSeatByPlayer(cmd.PlayerRoot) >= 0 {
		return 0, angzarr.NewCommandRejectedError("Player already seated")
	}

	if cmd.BuyInAmount < state.MinBuyIn {
		return 0, angzarr.NewCommandRejectedError(fmt.Sprintf("Buy-in must be at least %d", state.MinBuyIn))
	}
	if cmd.BuyInAmount > state.MaxBuyIn {
		return 0, angzarr.NewCommandRejectedError("Buy-in above maximum")
	}

	var seatPosition int32
	if cmd.PreferredSeat >= 0 && cmd.PreferredSeat < state.MaxPlayers {
		if _, taken := state.Seats[cmd.PreferredSeat]; taken {
			return 0, angzarr.NewCommandRejectedError("Seat is occupied")
		}
		seatPosition = cmd.PreferredSeat
	} else {
		seatPosition = state.NextAvailableSeat()
		if seatPosition < 0 {
			return 0, angzarr.NewCommandRejectedError("Table is full")
		}
	}

	return seatPosition, nil
}

func computePlayerJoined(cmd *examples.JoinTable, seatPosition int32) *examples.PlayerJoined {
	return &examples.PlayerJoined{
		PlayerRoot:   cmd.PlayerRoot,
		SeatPosition: seatPosition,
		BuyInAmount:  cmd.BuyInAmount,
		Stack:        cmd.BuyInAmount,
		JoinedAt:     timestamppb.New(time.Now()),
	}
}

// HandleJoinTable handles the JoinTable command.
func HandleJoinTable(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state TableState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.JoinTable
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardJoinTable(state); err != nil {
		return nil, err
	}
	seatPosition, err := validateJoinTable(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computePlayerJoined(&cmd, seatPosition)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}

File: examples/rust/table/agg/src/handlers/join.rs
//! JoinTable command handler.

use angzarr_client::proto::examples::{JoinTable, PlayerJoined};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::TableState;

pub fn handle_join_table(
    command_book: &CommandBook,
    command_any: &Any,
    state: &TableState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Table does not exist"));
    }

    let cmd: JoinTable = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    if cmd.player_root.is_empty() {
        return Err(CommandRejectedError::new("player_root is required"));
    }

    // Check if player already seated
    if state.find_seat_by_player(&cmd.player_root).is_some() {
        return Err(CommandRejectedError::new("Player already seated"));
    }

    // Validate buy-in
    if cmd.buy_in_amount < state.min_buy_in {
        return Err(CommandRejectedError::new(format!(
            "Buy-in must be at least {}",
            state.min_buy_in
        )));
    }
    if cmd.buy_in_amount > state.max_buy_in {
        return Err(CommandRejectedError::new("Buy-in above maximum"));
    }

    // Find seat
    let seat_position = if cmd.preferred_seat >= 0 && cmd.preferred_seat < state.max_players {
        if state.seats.contains_key(&cmd.preferred_seat) {
            return Err(CommandRejectedError::new("Seat is occupied"));
        }
        cmd.preferred_seat
    } else {
        state
            .next_available_seat()
            .ok_or_else(|| CommandRejectedError::new("Table is full"))?
    };

    let event = PlayerJoined {
        player_root: cmd.player_root,
        seat_position,
        buy_in_amount: cmd.buy_in_amount,
        stack: cmd.buy_in_amount,
        joined_at: Some(angzarr_client::now()),
    };

    let event_any = pack_event(&event, "examples.PlayerJoined");

    Ok(new_event_book(command_book, seq, event_any))
}
